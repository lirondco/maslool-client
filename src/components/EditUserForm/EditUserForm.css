.user_info button {
    margin: 0 !important;
}

.user_info {
    margin: 2%;
}

.user_info span {
    font-weight: bolder;
    font-family: monospace;
    font-size: x-large;
}

.Edit_button {
    text-decoration: underline;
    border: none;
    font-size: large;
    color: blue;
}

.Edit_button:hover {
    text-decoration: underline;
    border: none;
    font-size: large;
    color: purple;
}

.user_info form,
.user_info label,
.user_info input {
    margin: 0 !important;
    padding: 0 !important;
}

.user_info form {
    display: flex;
    flex-direction: row;
}

.user_info form button {
    border: none;
    font-weight: bold;
    text-decoration: underline;
    color: blue;
}

.user_info input {
    font-weight: bolder;
    font-family: monospace;
    border-radius: 0 !important;
}

@media screen and (min-width: 500px) {
.user_info label {
    margin-left: -25.7% !important;
}
}

.user_info form button:hover {
    border: none;
    font-weight: bold;
    text-decoration: underline;
    color: purple;
}