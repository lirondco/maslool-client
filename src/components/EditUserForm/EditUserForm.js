import React, { Component } from "react";
import UserContext from "../../contexts/UserContext";
import UsersApiService from "../../services/users-api-service";
import { Input, Label, NiceDate } from "../Utils/Utils";
import "./EditUserForm.css";

export default class EditUserForm extends Component {
  state = {
    user: null,
    isEditing: false,
  };

  static contextType = UserContext;

  componentDidMount() {
    const { user } = this.context;
    UsersApiService.getUser(user.id)
      .then((res) => this.setState({ user: res }))
      .catch(this.context.setError);
  }

  handleEditClick = (e) => {
    e.preventDefault();
    this.setState({ isEditing: true });
  };

  handleCancelClick = (e) => {
    e.preventDefault();
    this.setState({ isEditing: false });
  };

  handleSubmit = (ev) => {
    this.context.clearError();
    ev.preventDefault();
    const { user } = this.context;
    const { email } = ev.target;
    UsersApiService.editUser(user.id, email.value)
      .then(this.setState({ isEditing: false }))
      .then(() => (email.value = ""))
      .catch(this.context.setError);
    UsersApiService.getUser(user.id)
      .then((res) => this.setState({ user: res }))
      .catch(this.context.setError);
  };

  renderEditInterface = () => {
    return (
      <form type="submit" className="edit_email" onSubmit={this.handleSubmit}>
        <Label htmlFor="email">Email: </Label>
        <Input name="email" id="email" defaultValue={this.state.user.email} />
        <button type="submit">Change</button>
        <button onClick={this.handleCancelClick}>Cancel</button>
      </form>
    );
  };

  renderUserInfo = () => {
    return (
      <>
        <p>
          Username: <span>{this.state.user.username}</span>
        </p>
        {this.state.isEditing ? (
          this.renderEditInterface()
        ) : (
          <p>
            Email: <span>{this.state.user.email} </span>
            <button
              onClick={this.handleEditClick}
              type="button"
              className="Edit_button"
            >
              [edit]
            </button>
          </p>
        )}
        <p>
          Member since:{" "}
          <span>
            <NiceDate date={this.state.user.join_date} />
          </span>
        </p>
      </>
    );
  };

  render() {
    return (
      <div className="user_info">
        {this.state.user ? this.renderUserInfo() : <p>Loading ...</p>}
        <div role="alert">
          {this.context.error && (
            <p className="error">{this.context.error.error}</p>
          )}
        </div>
      </div>
    );
  }
}
