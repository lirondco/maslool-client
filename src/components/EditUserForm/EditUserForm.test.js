import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import EditUserForm from "./EditUserForm";
import { withContext } from "shallow-with-context";

describe(`EditUserForm component`, () => {
  const context = {
    clearError: () => {},
    user: { id: 3, admin: false, banned: false, username: "test_user" },
  };
  const ComponentWithContext = withContext(EditUserForm, context);
  
  it("renders the form", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    wrapper.setState({ isEditing: true, user: {id: 3, admin: false, banned: false, username: "test_user", email: "test@ema.il"}})
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("form submits when submitted", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    wrapper.setState({ isEditing: true, user: {id: 3, admin: false, banned: false, username: "test_user", email: "test@ema.il"}})
    const mockedEvent = {
      preventDefault: () => {},
      target: { email: { value: 'test@ema.il' } }
    };
    wrapper.find("form").at(0).simulate("submit", mockedEvent);
  });
});
