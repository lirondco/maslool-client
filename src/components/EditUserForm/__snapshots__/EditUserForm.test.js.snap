// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserForm component renders the form 1`] = `
<div
  className="user_info"
>
  <p>
    Username: 
    <span>
      test_user
    </span>
  </p>
  <form
    className="edit_email"
    onSubmit={[Function]}
    type="submit"
  >
    <Label
      htmlFor="email"
    >
      Email: 
    </Label>
    <ForwardRef
      defaultValue="test@ema.il"
      id="email"
      name="email"
    />
    <button
      type="submit"
    >
      Change
    </button>
    <button
      onClick={[Function]}
    >
      Cancel
    </button>
  </form>
  <p>
    Member since:
     
    <span>
      <NiceDate />
    </span>
  </p>
  <div
    role="alert"
  />
</div>
`;
